native/explorer/src/lazyframe.rs

```diff
@@ -0,0 +1,71 @@
+use polars::prelude::*;
+use std::result::Result;
+
+use crate::{ExDataFrame, ExLazyFrame, ExSeries, ExplorerError};
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_collect(data: ExLazyFrame) -> Result<ExDataFrame, ExplorerError> {
+    Ok(ExDataFrame::new(data.resource.0.clone().collect()?))
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_fetch(data: ExLazyFrame, n_rows: usize) -> Result<ExDataFrame, ExplorerError> {
+    Ok(ExDataFrame::new(data.resource.0.clone().fetch(n_rows)?))
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_describe_plan(data: ExLazyFrame, optimized: bool) -> Result<String, ExplorerError> {
+    let lf = &data.resource.0;
+    let plan = match optimized {
+        true => lf.describe_optimized_plan()?,
+        false => lf.describe_plan(),
+    };
+    Ok(plan)
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_head(data: ExLazyFrame, length: u32) -> Result<ExLazyFrame, ExplorerError> {
+    let lf = &data.resource.0;
+    Ok(ExLazyFrame::new(lf.clone().limit(length)))
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_tail(data: ExLazyFrame, length: u32) -> Result<ExLazyFrame, ExplorerError> {
+    let lf = &data.resource.0;
+    Ok(ExLazyFrame::new(lf.clone().tail(length)))
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_pull(data: ExLazyFrame, name: &str) -> Result<ExSeries, ExplorerError> {
+    let df = &data.resource.0.clone().select(&[col(name)]).collect()?;
+    let series = df.column(name).map(|s| ExSeries::new(s.clone()))?;
+    Ok(series)
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_names(data: ExLazyFrame) -> Result<Vec<String>, ExplorerError> {
+    let lf = &data.resource.0;
+    Ok(lf.schema().iter_names().cloned().collect())
+}
+
+#[rustler::nif(schedule = "DirtyCpu")]
+pub fn lf_dtypes(data: ExLazyFrame) -> Result<Vec<String>, ExplorerError> {
+    let lf = &data.resource.0;
+    Ok(lf
+        .schema()
+        .iter_dtypes()
+        .map(|dtype| dtype.to_string())
+        .collect())
```

Comment:
I have a question: how expensive are the names and dtypes operations? Does LazyPolars always keep it readily available? Or does it actually have to execute some of the computations?

Reply:
Fairly inexpensive. They use the logical plan, which has to be built, but is fairly instantaneous.

Reply:
Ok, one of the things I was talking to @philss is that we could keep the computation of names and types within `Explorer.DataFrame` (like we keep `groups`). I think ExplorerSQL would need to keep track of names and types _without_ realizing the plan (which is fairly more expensive there)? So we can either:

1. Implement this in ExplorerSQL
2. Implement this in Explorer.DataFrame so it benefits everyone (including LazyPolars)

Thoughts?

Reply:
Btw, we don't need to decide it for this PR, but something we will have to sync fairly soon!

Reply:
Let's sync on this soon and figure it out. It's not setting alarm bells ringing for me except for keeping the names/dtypes in order. Given it's an instant lookup in eager polars and very cheap in lazy we could even just call the NIF on every `update_dataframe` call, which would be easy enough.